src/riscv/linux/riscv-hw.c

```diff
@@ -1,10 +1,15 @@
-#include <sys/hwprobe.h>
 #include <sched.h>
-
 #include <cpuinfo/log.h>
+
 #include <riscv/api.h>
 #include <riscv/linux/api.h>
 
+#include <linux/version.h>
+
+#if LINUX_VERSION_CODE >= KERNEL_VERSION(6,4,0)
```

Comment:
So . . . .philosophical question to @malfet @GlassOfWhiskey . I know current hardware is not all at 6.4 (I guess, to be more fair, most of it isn't), but hwprobe is supposed to be the "proper" way of querying this information. I think it's fair to say that no one likes the prospect of trying to string-parse a proc-file to infer data.

This actually goes back to my comment on the other file about whether we should take this opportunity to just remove all the features that aren't useful to detect (e.g. why do we want to detect 'm', literally everyone will have 'm'). If we do it that way, then the question becomes "if you're only here to detect, say "V", well V isn't available till Linux 6.5, so you *must* have hwprobe". What do you both think?

Reply:
This #ifdef does not fix the RISC-V build on machines with a 6.4 kernel or greater.  When compiling this branch on a RISC-V 64 Ubuntu 23.10 VM, which has a 6.5 kernel, I get the following errors

```
cpuinfo/src/riscv/linux/riscv-hw.c:11:10: fatal error: sys/hwprobe.h: No such file or directory
   11 | #include <sys/hwprobe.h>
      |          ^~~~~~~~~~~~~~~
compilation terminated.
```
It's not clear to me where sys/hwprobe.h is supposed to come from.  @prashanthswami I think you added this originally.  Were you testing with your own build of glibc?

Also, IIUC, with this compile time check on the kernel version, hwprobe will only be used if both the build machine and the system running the binary have a 6.4 kernel or greater.  If we build on Ubuntu 22.04 for example but run on 23.10 the hwprobe path would not be taken here even though 23.10 has a 6.5 kernel with hwprobe support.

Other projects that are using hwprobe right now get around this problem by making the syscall directly. [OpenJDK](https://github.com/openjdk/jdk/blob/master/src/hotspot/os_cpu/linux_riscv/riscv_hwprobe.cpp) duplicates the constants from the kernel headers needed for the sys call and then does a direct syscall. It doesn't include the kernel header files at all.  A nicer solution would be to first check to see if the <asm/hwprobe.h> exists either using a build system check that defines a macro if a program that includes <asm/hwprobe.h> builds or perhaps using the trick the glibc patch uses

```
#ifdef __has_include
# if __has_include (<asm/hwprobe.h>)
#  include <asm/hwprobe.h>
# endif
#endif
```

Regardless of the method used, if the include file is not found,  manually defined constants for the hwprobe syscall number and the various bitmasks can be defined, so the syscall can still be issued, regardless of the kernel version on the build machine.

Reply:
My primary use-case was [Android on RISC-V](https://github.com/google/android-riscv64), so I likely overcorrected here by testing primarily against the Android NDK (note, there's no official NDK with RISC-V support yet, but the public canary builds work). This header is available in [bionic](https://cs.android.com/android/platform/superproject/main/+/main:bionic/libc/kernel/uapi/asm-riscv/asm/hwprobe.h), so building against the NDK does not complain.

The proposal to check for the header explicitly seems reasonable enough to me and there's precedent for it.

As a side note - we should indeed have a CI bot for this. I can't add one for Android just yet, because the common Github workflows don't support canary Android NDKs, but I've filed #206 to do so when it is. Separately, it would probably be valuable to run a linux-riscv64 build to catch stuff like this, so I'll figure out that out.

Reply:
Ah, that explains it.  It looks like you're pulling in this [header](https://cs.android.com/android/platform/superproject/main/+/main:bionic/libc/include/sys/hwprobe.h) which contains a definition of __riscv_hwprobe that appears to be compatible with the pending glibc patch.  Ultimately, the thing to do would be to check for <sys/hwprobe.h> and, if it's not present, fall back to code that makes the syscall directly.  However, perhaps until the glibc patch is merged, <sys/hwprobe.h> should only be included when building for Android.  A manual riscv_hwprobe syscall could then be used for other Linux distributions for the time being.

Reply:
@malfet, @GlassOfWhiskey , @prashanthswami The build issues on mainline Linux distributions are still not addressed by this patch.  It hasn't been possible to build pytorch/cpuinfo on Ubuntu 23.10 for over 3 weeks and this patch in it's current form won't help.  Would there by any objections from anyone if I submit a PR to fix the build issues?  The PR would only address the build issues to avoid introducing too many conflicts.  Other issues with hwprobe, like the fact we're calling it far too often in most cases, could be addressed in a subsequent PR.

Reply:
@markdryan no problem from my side. Also because probably things will go a bit longer in approving this PR as many aspects are involved.
I missed what was the preferred solution among the ones proposed so I didn't do anything on that direction for now. Feel free to proceed as you prefer.
The `__has_include` path is probably the most elegant one. It is not fully C standard, but it is now part of the C++ standard since C++17 so I'd vote for that one.